File: solutions/javascript/ShuffleanArray/ShuffleanArray.js
/**
 * @param {number[]} nums
 */
var Solution = function(nums) {
    this.origin = nums.slice();
    this.arr = nums;
};

/**
 * Resets the array to its original configuration and return it.
 * @return {number[]}
 */
Solution.prototype.reset = function() {
    return this.origin;
};

/**
 * Returns a random shuffling of the array.
 * @return {number[]}
 */
Solution.prototype.shuffle = function() {
    var xxx = this.arr;
    for(var j,x,i=xxx.length;i;j=parseInt(Math.random()*i),x=xxx[--i],xxx[i]=xxx[j],xxx[j]=x);
    return xxx;
};

/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = Object.create(Solution).createNew(nums)
 * var param_1 = obj.reset()
 * var param_2 = obj.shuffle()
 */

File: solutions/java/ShuffleanArray/ShuffleanArray.java
class Solution {

    private int[] origin;
    private Random rand;

    public Solution(int[] nums) {
        origin = nums;
        rand = new Random();
    }

    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return origin;
    }

    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        int[] arr = origin.clone();
        for (int i = arr.length - 1; i > 0; i--) {
            int rd = rand.nextInt(i + 1);
            int tmp = arr[i];
            arr[i] = arr[rd];
            arr[rd] = tmp;
        }
        return arr;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */